engine/src/lib.rs

```diff
@@ -69,20 +62,66 @@ impl ServalEngine {
     /// Run the passed-in WASM executable on the given input bytes.
     pub fn execute(
         &mut self,
+        // WebAssembly binary to execute
         binary: &[u8],
+        // Data to pass to WebAssembly as stdin
         input: &[u8],
+        // List of elevated permissions for this binary
+        permissions: &[Permission],
     ) -> Result<WasmResult, ServalEngineError> {
         let stdout = WritePipe::new_in_memory();
         let stderr = WritePipe::new_in_memory();
 
+        // Link the experimental HTTP support
+        let allowed_http_hosts = if permissions.contains(&Permission::AllHttpHosts) {
+            // todo: unclear whether we should actually support a wildcard like this
+            vec!["insecure:allow-all".to_string()]
+        } else {
+            permissions
+                .iter()
+                .filter_map(|perm| match perm {
+                    Permission::HttpHost(origin) => Some(origin.to_owned()),
+                    _ => None,
+                })
+                .collect()
+        };
+
+        if !allowed_http_hosts.is_empty() {
+            let http_state =
+                HttpState::new().map_err(ServalEngineError::EngineInitializationError)?;
+            http_state
+                .add_to_linker(&mut self.linker, move |_| HttpCtx {
+                    // todo: there's some confusion around whether
+                    allowed_hosts: Some(allowed_http_hosts.clone()),
+                    max_concurrent_requests: Some(42),
+                })
+                .map_err(ServalEngineError::EngineInitializationError)?;
+        }
+
         let stdin = ReadPipe::from(input);
-        let wasi = WasiCtxBuilder::new()
+        let mut wasi_builder = WasiCtxBuilder::new()
             .stdin(Box::new(stdin))
             .stdout(Box::new(stdout.clone()))
-            .stderr(Box::new(stderr.clone()))
-            .build();
+            .stderr(Box::new(stderr.clone()));
+
+        // Give the engine access to whichever parts of the file system are required
+        // TODO: this list should be pulled from the job's manifest, and permissions should be
+        // checked against the owner of the job in question and the configuration of this node (that
+        // is, the list of file system locations that any job by any user can access should be
+        // defined at the node level, and then the subset of those that any job by a specific user
+        // can access should exist in our configuration store, and then the subset of those that a
+        // specific job by that specific user should exist in the manifest for that job. Phew!
+        if permissions.contains(&Permission::ProcRead) {
+            let path = PathBuf::from("/proc");
```

Comment:
Would it make sense to test for `std::env::consts::OS == "linux"` (or a better test for whether we're running on linux) before even attempting this, in order to return a more helpful error? This could be confusing if proc access is explicitly pointed out as a feature and then returns "unsupported feature". It sure is unsupported, but for OS reasons specific to the host it is running on, not because the overall runtime doesn't support it yet.

Reply:
Fun fact: no. 😂 At first, I implemented this using `#[cfg(target_os = "linux")]` and it ended up leading to (incorrect) dead code warnings whenever I compiled it on macOS. 🤦